File: apps/api/src/networks/networks.service.ts
import { HttpStatus, Inject, Injectable } from '@nestjs/common';
import { Network } from '@prisma/client';
import {
  NetworkDetail,
  NetworkSummary,
  WS_EVENTS,
} from '@nodescope/shared';
import { AuditService } from '../audit/audit.service';
import { NodeScopeException } from '../common/filters/global-exception.filter';
import { ConflictResolutionService } from '../conflict/conflict.service';
import type { OrgMemberContext } from '../organizations/org-context.types';
import { PermissionsService } from '../permissions/permissions.service';
import { IRealtimeService, REALTIME_SERVICE } from '../realtime/realtime.types';
import {
  CreateNetworkDto,
  NETWORK_WRITABLE_FIELDS,
  PatchNetworkDto,
} from './networks.dto';
import { NetworksRepository } from './networks.repository';

const MAX_NETWORKS_PER_ORG = 1;

@Injectable()
export class NetworksService {
  constructor(
    private readonly networksRepository: NetworksRepository,
    private readonly conflictService: ConflictResolutionService,
    @Inject(REALTIME_SERVICE) private readonly realtimeService: IRealtimeService,
    private readonly audit: AuditService,
    private readonly permissions: PermissionsService,
  ) {}

  async listNetworks(member: OrgMemberContext): Promise<NetworkSummary[]> {
    const scope = await this.permissions.scopeFilter(member);
    const networks = await this.networksRepository.listVisible(member.organizationId, scope);
    return networks.map((n) => this.toSummary(n));
  }

  async createNetwork(
    member: OrgMemberContext,
    creatorUserId: string,
    dto: CreateNetworkDto,
  ): Promise<NetworkDetail> {
    // No charters yet on create — just the role gate
    await this.permissions.assertNetworkFullCoverage(member, []);
    const organizationId = member.organizationId;

    const existingCount = await this.networksRepository.countByOrgId(organizationId);
    if (existingCount >= MAX_NETWORKS_PER_ORG) {
      throw new NodeScopeException(
        'NETWORK_001',
        'NETWORK_LIMIT_EXCEEDED',
        HttpStatus.CONFLICT,
      );
    }

    const network = await this.networksRepository.create({
      organizationId,
      userId: creatorUserId,
      ...dto,
    });
    await this.audit.recordCreate(organizationId, 'Network', network);
    const detail = this.toDetail(network);
    this.conflictService.emitEntityEvent(
      WS_EVENTS.NETWORK_UPDATED,
      { networkId: network.id, network: detail },
      organizationId,
    );
    return detail;
  }

  async getNetwork(member: OrgMemberContext, networkId: string): Promise<NetworkDetail> {
    const scope = await this.permissions.scopeFilter(member);
    const network = await this.networksRepository.findVisibleByIdAndOrgId(
      networkId,
      member.organizationId,
      scope,
    );
    if (!network) {
      throw new NodeScopeException('NETWORK_002', 'NETWORK_NOT_FOUND', HttpStatus.NOT_FOUND);
    }
    return this.toDetail(network);
  }

  async updateNetwork(
    member: OrgMemberContext,
    networkId: string,
    patch: PatchNetworkDto,
  ): Promise<NetworkDetail> {
    const organizationId = member.organizationId;
    // Write path: org-wide lookup so out-of-scope ADMIN → PERM_004 not 404
    const network = await this.networksRepository.findByIdAndOrgId(networkId, organizationId);
    if (!network) {
      throw new NodeScopeException('NETWORK_002', 'NETWORK_NOT_FOUND', HttpStatus.NOT_FOUND);
    }

    const charteredIds = await this.networksRepository.charteredPropertyIds(organizationId, networkId);
    await this.permissions.assertNetworkFullCoverage(member, charteredIds);

    const updatePayload = this.conflictService.buildUpdatePayload(
      patch,
      NETWORK_WRITABLE_FIELDS,
      network.version,
      CreateNetworkDto,
    );

    const updated = await this.networksRepository.updateWithVersion(
      networkId,
      organizationId,
      updatePayload,
      patch.baseVersion,
    );
    if (!updated) {
      throw new NodeScopeException('SYNC_001', 'EDIT_CONFLICT', HttpStatus.CONFLICT);
    }

    const detail = this.toDetail(updated);
    await this.audit.recordUpdate(organizationId, 'Network', networkId, patch.changes);
    this.conflictService.emitEntityEvent(
      WS_EVENTS.NETWORK_UPDATED,
      { networkId, network: detail, changes: patch.changes, updatedBy: updated.userId ?? '' },
      organizationId,
    );

    if (patch.changes.some((c) => c.field === 'homePublicIp')) {
      void this.realtimeService.recomputeOnHomeForUser(updated.userId ?? '');
    }

    return detail;
  }

  async setHomeIpFromRequest(
    member: OrgMemberContext,
    networkId: string,
    requestIp: string,
  ): Promise<NetworkDetail> {
    // Org-wide lookup to get the version for the changeset (coverage check runs inside updateNetwork)
    const organizationId = member.organizationId;
    const network = await this.networksRepository.findByIdAndOrgId(networkId, organizationId);
    if (!network) {
      throw new NodeScopeException('NETWORK_002', 'NETWORK_NOT_FOUND', HttpStatus.NOT_FOUND);
    }
    return this.updateNetwork(member, networkId, {
      baseVersion: network.version,
      changes: [
        {
          field: 'homePublicIp',
          oldValue: network.homePublicIp,
          newValue: requestIp,
        },
      ],
    });
  }

  async deleteNetwork(member: OrgMemberContext, networkId: string): Promise<void> {
    const organizationId = member.organizationId;
    // Write path: org-wide lookup so out-of-scope ADMIN → PERM_004 not 404
    const network = await this.networksRepository.findByIdAndOrgId(networkId, organizationId);
    if (!network) {
      throw new NodeScopeException('NETWORK_002', 'NETWORK_NOT_FOUND', HttpStatus.NOT_FOUND);
    }
    const charteredIds = await this.networksRepository.charteredPropertyIds(organizationId, networkId);
    await this.permissions.assertNetworkFullCoverage(member, charteredIds);
    await this.networksRepository.deleteByIdAndOrgId(networkId, organizationId);
    await this.audit.recordDelete(organizationId, 'Network', network);
  }

  /**
   * Called by the realtime gateway (which only knows userId, not organizationId).
   * Looks up the user's org's first network via org membership and checks whether
   * the request IP matches the network's homePublicIp.
   */
  async checkOnHome(
    userId: string,
    requestIp: string,
  ): Promise<{ networkId: string | null; onHome: boolean }> {
    const networks = await this.networksRepository.findAllByMemberUserId(userId);
    const network = networks[0] ?? null;
    if (!network) return { networkId: null, onHome: false };

    const onHome =
      requestIp.length > 0 &&
      network.homePublicIp !== null &&
      network.homePublicIp === requestIp;
    return { networkId: network.id, onHome };
  }

  private toSummary(network: Network): NetworkSummary {
    return {
      id: network.id,
      name: network.name,
      homeAddress: network.homeAddress,
      homeLatitude: network.homeLatitude,
      homeLongitude: network.homeLongitude,
      isp: network.isp,
      downMbps: network.downMbps,
      upMbps: network.upMbps,
      version: network.version,
      createdAt: network.createdAt.toISOString(),
      updatedAt: network.updatedAt.toISOString(),
    };
  }

  private toDetail(network: Network): NetworkDetail {
    return { ...this.toSummary(network), homePublicIp: network.homePublicIp };
  }
}

File: src/Modules/Inventory/NodeScope.Modules.Inventory.Application/Networks/NetworksService.cs
using System.Text.Json;
using NodeScope.Contracts.Realtime;
using NodeScope.Platform.Abstractions;

namespace NodeScope.Modules.Inventory.Application.Networks;

/// <summary>Network CRUD, ported from Node's <c>NetworksService</c>.</summary>
public sealed class NetworksService
{
    private const int MaxNetworksPerOrg = 1;

    private readonly INetworkRepository _networks;
    private readonly IPermissionScopeService _permissions;
    private readonly IRealtimeService _realtime;
    private readonly IAuditService _audit;

    public NetworksService(
        INetworkRepository networks,
        IPermissionScopeService permissions,
        IRealtimeService realtime,
        IAuditService audit)
    {
        _networks = networks;
        _permissions = permissions;
        _realtime = realtime;
        _audit = audit;
    }

    public async Task<IReadOnlyList<NetworkSummaryDto>> ListAsync(
        OrgMemberContext member,
        CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(member);
        var scope = await _permissions.ScopePropertyIdsAsync(member, cancellationToken);
        var networks = await _networks.ListVisibleAsync(member.OrganizationId, scope, cancellationToken);
        return [.. networks.Select(network => network.ToSummary())];
    }

    public async Task<NetworkDetailDto> CreateAsync(
        OrgMemberContext member,
        string creatorUserId,
        CreateNetworkRequest request,
        CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(member);
        ArgumentNullException.ThrowIfNull(request);
        // No charters yet on create - just the role gate.
        await _permissions.AssertNetworkFullCoverageAsync(member, [], cancellationToken);
        var organizationId = member.OrganizationId;

        if (await _networks.CountByOrgAsync(organizationId, cancellationToken) >= MaxNetworksPerOrg)
        {
            throw InventoryErrors.NetworkLimitExceeded();
        }

        var network = await _networks.CreateAsync(
            new NewNetwork(
                organizationId,
                creatorUserId,
                request.TrimmedName!,
                request.HomeAddress,
                request.HomeLatitude,
                request.HomeLongitude,
                request.HomePublicIp,
                request.Isp,
                request.DownMbps,
                request.UpMbps),
            cancellationToken);
        var detail = network.ToDetail();
        await _audit.RecordCreateAsync(organizationId, "Network", network.Id, detail, cancellationToken);
        // No charters yet on create means an OWNER-only fan-out.
        await _realtime.EmitScopedMultiAsync(
            organizationId,
            [],
            WsEvents.NetworkUpdated,
            new { networkId = network.Id, network = detail, timestamp = IsoTimestamp.Now() },
            cancellationToken);
        return detail;
    }

    public async Task<NetworkDetailDto> GetAsync(
        OrgMemberContext member,
        string networkId,
        CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(member);
        var scope = await _permissions.ScopePropertyIdsAsync(member, cancellationToken);
        var network = await _networks.FindVisibleAsync(member.OrganizationId, networkId, scope, cancellationToken)
            ?? throw InventoryErrors.NetworkNotFound();
        return network.ToDetail();
    }

    public async Task<NetworkDetailDto> UpdateAsync(
        OrgMemberContext member,
        string networkId,
        ChangesetRequest patch,
        CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(member);
        ArgumentNullException.ThrowIfNull(patch);
        var organizationId = member.OrganizationId;
        // Write path: org-wide lookup so an out-of-scope ADMIN gets PERM_004, not 404.
        var network = await _networks.FindAsync(organizationId, networkId, cancellationToken)
            ?? throw InventoryErrors.NetworkNotFound();
        await _permissions.AssertNetworkFullCoverageAsync(
            member, await CoverageSitesAsync(organizationId, networkId, cancellationToken), cancellationToken);

        var payload = Changesets.BuildUpdatePayload(patch, NetworkFields.Writable, network.Version, NetworkFields.IsValid);
        var updated = await _networks.UpdateWithVersionAsync(
            organizationId, networkId, payload, patch.BaseVersion!.Value, cancellationToken)
            ?? throw Changesets.EditConflict();

        var detail = updated.ToDetail();
        await _audit.RecordUpdateAsync(
            organizationId, "Network", networkId, ChangesetAudit.ToFieldChanges(patch), cancellationToken);
        var charterSites = await _networks.CharteredPropertyIdsAsync(organizationId, networkId, cancellationToken);
        await _realtime.EmitScopedMultiAsync(
            organizationId,
            charterSites,
            WsEvents.NetworkUpdated,
            new
            {
                networkId,
                network = detail,
                changes = patch.Changes,
                updatedBy = updated.UserId ?? "",
                timestamp = IsoTimestamp.Now(),
            },
            cancellationToken);

        if (patch.Changes!.Any(change => change.Field == "homePublicIp"))
        {
            await _realtime.RecomputeOnHomeForUserAsync(updated.UserId ?? "", cancellationToken);
        }

        return detail;
    }

    public async Task<NetworkDetailDto> SetHomeIpAsync(
        OrgMemberContext member,
        string networkId,
        string requestIp,
        CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(member);
        var network = await _networks.FindAsync(member.OrganizationId, networkId, cancellationToken)
            ?? throw InventoryErrors.NetworkNotFound();
        using var newValue = JsonDocument.Parse(JsonSerializer.Serialize(requestIp));
        using var oldValue = JsonDocument.Parse(JsonSerializer.Serialize(network.HomePublicIp));
        return await UpdateAsync(
            member,
            networkId,
            new ChangesetRequest
            {
                BaseVersion = network.Version,
                Changes =
                [
                    new ChangesetChange
                    {
                        Field = "homePublicIp",
                        OldValue = oldValue.RootElement.Clone(),
                        NewValue = newValue.RootElement.Clone(),
                    },
                ],
            },
            cancellationToken);
    }

    public async Task DeleteAsync(OrgMemberContext member, string networkId, CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(member);
        var organizationId = member.OrganizationId;
        var network = await _networks.FindAsync(organizationId, networkId, cancellationToken)
            ?? throw InventoryErrors.NetworkNotFound();
        await _permissions.AssertNetworkFullCoverageAsync(
            member, await CoverageSitesAsync(organizationId, networkId, cancellationToken), cancellationToken);
        await _networks.DeleteAsync(organizationId, networkId, cancellationToken);
        await _audit.RecordDeleteAsync(organizationId, "Network", networkId, network.ToDetail(), cancellationToken);
    }

    private async Task<IReadOnlyList<string>> CoverageSitesAsync(
        string organizationId,
        string networkId,
        CancellationToken cancellationToken) =>
        (await _networks.CharteredPropertyIdsAsync(organizationId, networkId, cancellationToken))
            .Union(
                await _networks.DeviceFootprintPropertyIdsAsync(organizationId, networkId, cancellationToken),
                StringComparer.Ordinal)
            .ToList();
}
